File: python/infinicore/nn/functional/softshrink.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def softshrink(input, lambd=0.5, *, out=None):
    if out is None:
        return Tensor(_infinicore.softshrink(input._underlying, float(lambd)))

    _infinicore.softshrink_(out._underlying, input._underlying, float(lambd))

    return out

File: src/infinicore/ops/softshrink/softshrink.cc
#include "infinicore/ops/softshrink.hpp"

namespace infinicore::op {

common::OpDispatcher<Softshrink::schema> &Softshrink::dispatcher() {
    static common::OpDispatcher<Softshrink::schema> dispatcher_;
    return dispatcher_;
}

void Softshrink::execute(Tensor output, Tensor input, float lambda) {
    dispatcher().lookup(context::getDevice().getType())(output, input, lambda);
}

Tensor softshrink(Tensor input, float lambda) {
    auto output = Tensor::empty(input->shape(), input->dtype(), input->device());
    softshrink_(output, input, lambda);
    return output;
}

void softshrink_(Tensor output, Tensor input, float lambda) {
    Softshrink::execute(output, input, lambda);
}

} // namespace infinicore::op
